app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,181 @@
+package org.mozilla.focus.biometrics
+
+import android.arch.lifecycle.Lifecycle
+import android.arch.lifecycle.LifecycleObserver
+import android.arch.lifecycle.OnLifecycleEvent
+import android.arch.lifecycle.ProcessLifecycleOwner
+import android.content.Context
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyPermanentlyInvalidatedException
+import android.security.keystore.KeyProperties
+import android.support.annotation.RequiresApi
+import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
+import android.support.v4.os.CancellationSignal
+import java.io.IOException
+import java.security.*
+import java.security.cert.CertificateException
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.NoSuchPaddingException
+import javax.crypto.SecretKey
+
+@RequiresApi(Build.VERSION_CODES.M)
+class BiometricAuthenticationHandler(
+        private val context: Context) : FingerprintManagerCompat.AuthenticationCallback(), LifecycleObserver {
+
+    private var fingerprintManager = FingerprintManagerCompat.from(context)
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+    private var keyStore: KeyStore? = null
+    private var keyGenerator: KeyGenerator? = null
+    private var cryptoObject: FingerprintManagerCompat.CryptoObject? = null
+
+    var needsAuth = false; private set
+    var biometricFragment: BiometricAuthenticationDialogFragment? = null; private set
+
+    init {
+        try {
+            keyStore = KeyStore.getInstance("AndroidKeyStore")
+
+        } catch (err: KeyStoreException) {
+            throw RuntimeException("Failed to get instance of KeyStore", err)
+        }
+
+        try {
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchProviderException ->
+                    throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+                else -> throw err
+            }
+        }
+
+        createKey(DEFAULT_KEY_NAME, false)
+
+        val defaultCipher: Cipher
+        try {
+            defaultCipher = Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/"
+                    + KeyProperties.BLOCK_MODE_CBC + "/"
+                    + KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchPaddingException ->
+                    throw RuntimeException("Failed to get an instance of Cipher", err)
+                else -> throw err
```

Comment:
\[detekt\] ERROR: \[exceptions/RethrowCaughtException\] Do not rethrow a caught exception of the same type